### DISCUSSION TITLE
face swap inpaint help

### DISCUSSION THREAD

**Discussion by thomasbenjaminhardy** [lllyasviel/Fooocus#1376] [Q&A] [answered]
Hey,

I'm new and learning, I've been using Fooocus for about a week now. I'm experimenting with the face swap feature, lately I've been testing the inpaint workflow. I mask out the face, then in the image prompts I add an image of a face and set the mode to face swap.
Then in the advanced tab under control, check the mix image prompt and inpaint.

What I've noticed is that as the image process develops, the face outcome is quite close to the input face, but then at a point near the end it seems to switch away completely and changes the face- losing the likeness. 

Sample Gif:
![faceSwap_issue_001](https://github.com/lllyasviel/Fooocus/assets/47830599/d3f80801-ba28-41c8-a419-5812def87d96)

Am I missing something in my settings, is there anyone who's had success with similar workflows? 

Thanks,

**Comment by rivboyj** [comment]
I experienced the same thing when i had the refiner model turned on. Turning that off makes it a continuous generation and doesnt completely switch things at the end

**Reply by thomasbenjaminhardy** [reply]
You're right, turning it off removes that additional change - thanks!

**Comment by Phrixus2023** [comment, marked as answer]
You chose the wrong function, you should have chosen the improvement details in the inpaint method

**Reply by thomasbenjaminhardy** [reply]
Thanks, this mode does indeed work better! thanks

**Comment by thomasbenjaminhardy** [comment]
After some additional testing, this is the workflow that I found works pretty well.

1. Fresh UI open
2. Add face image to image prompt.
3. Set mode to face swap, set 'Stop At' to 1, and 'Weight' to 1.
4. Add image to Inpaint tab, set method to 'Improve Detail..." and mask the face area.
5. Go to the advanced tab, enable 'Developer Debug Mode' and enable 'Mixing Image Prompt and Inpaint'. 
6. Hit generate!

The results are much closer than before, but the original face structure is retained to an extent. As an additional step I take the output image into photoshop and using the liquify tool I nudge the proportions closer to the desired face. 

Once I have a proportional match I put it back through, the result is very close, I imagine doing passes this way could get you near to perfect results. 

![Process_001](https://github.com/lllyasviel/Fooocus/assets/47830599/9df6c461-fa39-491f-a573-fc83adf1d87a)


Hope this helps!

**Reply by lgwjames** [reply]
Im going to have a go now, as some one asked me to try a few things and wasn't getting anywhere

**Comment by TanvirHafiz** [comment]
How did you mask out the face? like can you mask out in the generation tab?

**Reply by thomasbenjaminhardy** [reply]
You paint the mask in the 'inpaint ' section - like you would normally.

**Reply by lgwjames** [reply]
im sorry to say your way didn't work for me on a new person i was trying to do, works great with my ai girl that i have done like 100 photos off.

on my work flow as above im doing this:

1.  load fooocus, this auto loads my neg prompt and 3 loras, perfect hands, eyes and a sd_xl_ example < not sure if i need this one
2. click input image and advance
3. under image prompt i add a file of the face like below post ( for my ai girl i use a master pic on all of the images )
4. click advance, faceswap and set to 8-9 and then from 800-938 to test
5. prompt is just a simple, a pretty female, age , hair type 
6. load any photo as a prompt into the in paint and mask out hair and face

i have tried inpaint or output and the settings in debug mode as per your guide, but the face is that comes out is nothing like the person i want.

example

imput
![WhatsApp Image 2024-02-07 at 17 25 10](https://github.com/lllyasviel/Fooocus/assets/35565479/73d0fa8d-2906-4718-b348-9e07992e8b42)
![WhatsApp Image 2024-02-08 at 18 19 06](https://github.com/lllyasviel/Fooocus/assets/35565479/f51ecaed-14e5-4dd9-9ab4-ebb4192ed59e)

output

![2024-02-10_12-30-20_5620](https://github.com/lllyasviel/Fooocus/assets/35565479/def773c3-a93e-42ea-9546-38fb235d4b48)

![WhatsApp Image 2024-02-07 at 21 51 36](https://github.com/lllyasviel/Fooocus/assets/35565479/23493084-4e67-45b6-8df1-f235ed015617)

i seem to get better resets just adding a photo and playing with the weight etc in faceswap
for me its a pain, as a AMD mac user takes 15-20 mins for a image so a lot of time to play with settings

**Comment by infinity0** [comment]
See #2644, Upscale works much better than Inpainting.
